Increasing test coverage rclcpp_lifecycle

brawner opened the pull request:
As with #1043 and #1044, this PR increases the test coverage for rclcpp_lifecycle. The test coverage increases from 61% to 100% of the `include` files and 87% of the `src` files (88% combined)

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9864)](http://ci.ros2.org/job/ci_linux/9864/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5480)](http://ci.ros2.org/job/ci_linux-aarch64/5480/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8024)](http://ci.ros2.org/job/ci_osx/8024/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9720)](http://ci.ros2.org/job/ci_windows/9720/)
* Coverage [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux_coverage&build=78)](https://ci.ros2.org/job/ci_linux_coverage/78/)

Depends on #1043, #1044
New commit: https://github.com/ros2/rclcpp/commit/7441e1da2454b1e82aa1bbf9a095df766f129382

ivanpauno commented on rclcpp_action/CMakeLists.txt:
```diff
@@ -98,6 +98,16 @@ if(BUILD_TESTING)
       ${PROJECT_NAME}
     )
   endif()
+
```
this PR also adds a lot of tests for `rclcpp_action`, were those added intentionally here?
Seems that the same tests were added in https://github.com/ros2/rclcpp/pull/1043

ivanpauno commented on rclcpp_components/CMakeLists.txt:
```diff
@@ -73,12 +73,21 @@ if(BUILD_TESTING)
+  set(components "${components}test_rclcpp_components::TestComponentThrows;$<TARGET_FILE:test_component>\n")
+
+  # Used to catch invalid resource entry
+  set(invalid_components "test_rclcpp_components::TestComponentFoo;;$<TARGET_FILE:test_component>\n")
```
Same about adding tests to `rclcpp_components`, sounds like a duplicate of https://github.com/ros2/rclcpp/pull/1044.

brawner commented:
@ivanpauno Because there was a lot of files added in these PRs, I broke them out into three different ones stacked on top of one another. There is no reason why they should be dependent on one another. I'll clean that up.

brawner commented:
Rebased on master and removed dependency from #1043 and #1044.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10564)](http://ci.ros2.org/job/ci_linux/10564/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6017)](http://ci.ros2.org/job/ci_linux-aarch64/6017/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8604)](http://ci.ros2.org/job/ci_osx/8604/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10446)](http://ci.ros2.org/job/ci_windows/10446/)

dirk-thomas commented on rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:
```diff
@@ -0,0 +1,346 @@
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
```
Calling `shutdown` from a different thread causes a race which was making this test fail in the nightly jobs since it has been added. See #1204 for the proposed fix.
